### PR TITLE
PROGRAMMES-5767 Add Promotion object & show on brand page

### PR DESCRIPTION
**Pull request by BPScott**
Add Amen Promotion object
Request promotions in TLEC controller
First pass at rendering promos on brand page

### CODE REVIEW

File: src/DsAmen/Organism/Promotion/PromotionPresenter.php [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    private $router;
+
+    /** @var Promotion */
+    private $promotion;
+
+    /** @var string[] */
+    private $actionIcon;
+
+    /** @var int */
+    private $duration;
```
Comment (tatitati) [inline src/DsAmen/Organism/Promotion/PromotionPresenter.php]: just wondering if we know the unit of this duration. ms?, secs?

Comment (tatitati) [inline src/DsAmen/Organism/Promotion/PromotionPresenter.php]: just wondering if we know the unit of this duration. ms?, secs?, minutes?

File: app/Resources/views/find_by_pid/tlec.html.twig [hunk 2]
```diff
@@ -28,6 +28,23 @@
...
+            {%- if promotions or clips or galleries -%}
+                <ul class="grid">
+                {%- for promotion in promotions -%}
+                    <li class="grid__item 1/2@gel3 1/3@gel3b 1/4@gel5">
+                        {# TODO Update image sizes for double width first promo #}
+                        {{- ds_amen('promotion', promotion, {
+                            media_variant: 'media--column media--card',
+                            cta_class: 'br-box-secondary',
+                            branding_name: 'secondary',
+                            link_location_prefix: 'promo_' ~ loop.index ~ '_',
```
Comment (votemike) [inline app/Resources/views/find_by_pid/tlec.html.twig]: Where is loop coming from?

File: app/Resources/views/find_by_pid/tlec.html.twig [hunk 3]
```diff
@@ -28,6 +28,23 @@
             {# TODO Recipes block #}
 
             {# TODO Promos / Clips / Galleries list #}
+            {%- if promotions or clips or galleries -%}
+                <ul class="grid">
+                {%- for promotion in promotions -%}
+                    <li class="grid__item 1/2@gel3 1/3@gel3b 1/4@gel5">
+                        {# TODO Update image sizes for double width first promo #}
+                        {{- ds_amen('promotion', promotion, {
```
Comment (votemike) [inline app/Resources/views/find_by_pid/tlec.html.twig]: This is a lot of options to have in a template. Would it be better to construct the presenter in the controller?

File: app/Resources/views/find_by_pid/tlec.html.twig [hunk 4]
```diff
@@ -28,6 +28,23 @@
             {# TODO Recipes block #}
 
             {# TODO Promos / Clips / Galleries list #}
+            {%- if promotions or clips or galleries -%}
```
Comment (votemike) [inline app/Resources/views/find_by_pid/tlec.html.twig]: Why clips or galleries? I can't see them being used.

File: src/DsAmen/Organism/Promotion/PromotionPresenter.php [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+class PromotionPresenter extends Presenter
+{
+    /** @var UrlGeneratorInterface */
+    private $router;
+
+    /** @var Promotion */
+    private $promotion;
+
+    /** @var string[] */
```
Comment (votemike) [inline src/DsAmen/Organism/Promotion/PromotionPresenter.php]: Does PHPStan complain about this being null to start with?

File: src/DsAmen/Organism/Promotion/PromotionPresenter.php [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+        return $this->promotion->getRelatedLinks();
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getActionIcon(): array
+    {
+        if ($this->actionIcon === null) {
```
Comment (votemike) [inline src/DsAmen/Organism/Promotion/PromotionPresenter.php]: Set in constructor instead?

File: src/DsAmen/Organism/Promotion/PromotionPresenter.php [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+                $this->duration = 0;
+            }
+        }
+
+        return $this->duration;
+    }
+
+    public function getBrandingBoxClass(): string
+    {
+        if (!empty($this->options['branding_name'])) {
```
Comment (votemike) [inline src/DsAmen/Organism/Promotion/PromotionPresenter.php]: Personal preference. Reverse this and similar checks?

File: src/DsAmen/Organism/Promotion/PromotionPresenter.php (hunk 1 above)
Comment (BPScott) [reply src/DsAmen/Organism/Promotion/PromotionPresenter.php]: Seconds. We don't do any transformation from the value we get back from PIPS.

File: tests/DsAmen/Organism/Promotion/PromotionPresenterTest.php [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+class PromotionPresenterTest extends TestCase
+{
+    private $mockRouter;
+
+    public function setUp()
+    {
+        $this->mockRouter = $this->createMock(UrlGeneratorInterface::class);
+    }
+
+    public function testPromotion()
```
Comment (tatitati) [inline tests/DsAmen/Organism/Promotion/PromotionPresenterTest.php]: can be more descriptive?, all this is about promotions. Maybe ```testPromotionReturnDefaultValues()``` ?

File: app/Resources/views/find_by_pid/tlec.html.twig (hunk 2 above)
Comment (BPScott) [reply app/Resources/views/find_by_pid/tlec.html.twig]: https://twig.symfony.com/doc/2.x/tags/for.html#the-loop-variable

File: app/Resources/views/find_by_pid/tlec.html.twig (hunk 4 above)
Comment (BPScott) [reply app/Resources/views/find_by_pid/tlec.html.twig]: To save people having to edit this line when we do add in the clips and galleries to this list. Possibly getting ahead of myself though :)

File: app/Resources/views/find_by_pid/tlec.html.twig (hunk 2 above)
Comment (votemike) [reply app/Resources/views/find_by_pid/tlec.html.twig]: YLSNED

File: tests/DsAmen/Organism/Promotion/PromotionPresenterTest.php [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+        $this->assertSame('short synopsis', $presenter->getSynopsis());
+        $this->assertSame($relatedLinks, $presenter->getRelatedLinks());
+        $this->assertSame(0, $presenter->getDuration());
+        $this->assertSame('br-box-subtle', $presenter->getBrandingBoxClass());
+        $this->assertSame('br-subtle-text-ontext', $presenter->getTextBrandingClass());
+    }
+
+    public function testPromotionOfProgrammeItem()
+    {
+        $this->mockRouter->method('generate')
```
Comment (tatitati) [inline tests/DsAmen/Organism/Promotion/PromotionPresenterTest.php]: The problem mixing expectations with assertions is that one doesn't know when is really expected this function be called. Is it on presenter->getImage(), presenter->getDuration(), any other?

File: app/Resources/views/find_by_pid/tlec.html.twig (hunk 3 above)
Comment (BPScott) [reply app/Resources/views/find_by_pid/tlec.html.twig]: On the existing brand page we create presenters in the controller and it ended up being a very ugly mess (as we never made a view helper for generating a promotion). 

These options are regarding how things look visually so IMO they belong in the view

File: src/DsAmen/Organism/Promotion/PromotionPresenter.php (hunk 5 above)
Comment (BPScott) [reply src/DsAmen/Organism/Promotion/PromotionPresenter.php]: It does not.

File: tests/DsAmen/Organism/Promotion/PromotionPresenterTest.php (hunk 9 above)
Comment (BPScott) [reply tests/DsAmen/Organism/Promotion/PromotionPresenterTest.php]: This is used when you call getUrl(). I'm not convinced it's worth creating a new test case just for that single function.

File: tests/DsAmen/Organism/Promotion/PromotionPresenterTest.php [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+        $this->assertSame('/programmes/b0000001', $presenter->getUrl());
+        $this->assertSame($image, $presenter->getImage());
+        // TV Episodes don't show a duration
+        $this->assertSame(0, $presenter->getDuration());
+        $this->assertSame(['set' => 'audio-visual', 'icon' => 'play'], $presenter->getActionIcon());
+    }
+
+    public function testPromotionOfImage()
+    {
+        $promotedEntity = $this->createMock(Image::class);
```
Comment (tatitati) [inline tests/DsAmen/Organism/Promotion/PromotionPresenterTest.php]: usually dummy objects are used to complete the signature of a function and carry on with the execution. Cause of that I wouldn't create it in a separate variable. That makes me think that is important enough to assert something later agains this variable, keeping it in a separate variable.

Comment (BPScott) [reply tests/DsAmen/Organism/Promotion/PromotionPresenterTest.php]: I do assert on the value. Scroll down a few lines:

```php
 $this->assertSame($image, $presenter->getImage());
```

File: src/DsAmen/Organism/Promotion/PromotionPresenter.php (hunk 5 above)
Comment (tatitati) [reply src/DsAmen/Organism/Promotion/PromotionPresenter.php]: simple :smile:

File: tests/DsAmen/Organism/Promotion/PromotionPresenterTest.php (hunk 10 above)
Comment (tatitati) [reply tests/DsAmen/Organism/Promotion/PromotionPresenterTest.php]: I'm fool. Sorry!

**Comment by programmes-developers** [comment]
Can one of the admins verify this patch?

**Comment by programmes-developers** [comment]
Can one of the admins verify this patch?

File: src/DsAmen/Organism/Promotion/PromotionPresenter.php (hunk 7 above)
Comment (BPScott) [reply src/DsAmen/Organism/Promotion/PromotionPresenter.php]: Usually i'd be with you, but in these cases I'm going for "when the data is absent early return the fallback" I think this reads a little better that mixing up if the fallback or the main case is within that first if statement.

File: src/DsAmen/Organism/Promotion/PromotionPresenter.php (hunk 6 above)
Comment (BPScott) [reply src/DsAmen/Organism/Promotion/PromotionPresenter.php]: Done. We should decide and write up a convention for this sort of ruling.

**Pull request merged**